libr/bin/dwarf.c

```diff
@@ -1263,104 +1335,155 @@ static const ut8 *r_bin_dwarf_parse_attr_value(const ut8 *obuf, int obuf_len,
 		}
 		break;
-	case DW_FORM_data2:
-		value->encoding.data = READ16 (buf);
-		break;
-	case DW_FORM_data4:
-		value->encoding.data = READ32 (buf);
-		break;
-	case DW_FORM_data8:
-		value->encoding.data = READ64 (buf);
-		break;
-	case DW_FORM_string:
-		value->encoding.str_struct.string = *buf? strdup ((const char*)buf) : NULL;
-		buf += (strlen ((const char*)buf) + 1);
-		break;
-	case DW_FORM_block:
+	case DW_FORM_block: // variable length ULEB128
 		buf = r_uleb128 (buf, buf_end - buf, &value->encoding.block.length);
 		if (!buf || buf >= buf_end) {
 			return NULL;
 		}
+		printf("%"PFMT64u" byte block:", value->encoding.block.length);
 		value->encoding.block.data = calloc (sizeof (ut8), value->encoding.block.length);
 		if (value->encoding.block.data) {
 			for (j = 0; j < value->encoding.block.length; j++) {
 				value->encoding.block.data[j] = READ (buf, ut8);
+				printf(" 0x%hhx", value->encoding.block.data[j]);
 			}
 		}
-		break;
-	case DW_FORM_block1:
-		value->encoding.block.length = READ (buf, ut8);
-		value->encoding.block.data = calloc (sizeof (ut8), value->encoding.block.length + 1);
-		if (value->encoding.block.data) {
-			for (j = 0; j < value->encoding.block.length; j++) {
-				value->encoding.block.data[j] = READ (buf, ut8);
-			}
-		}
+		printf("\n");
 		break;
 	case DW_FORM_flag:
 		value->encoding.flag = READ (buf, ut8);
-		break;
-	case DW_FORM_sdata:
-		buf = r_leb128 (buf, buf_end - buf, &value->encoding.sdata);
+		printf("%hhu\n", value->encoding.flag);
 		break;
 	case DW_FORM_strp:
-		value->encoding.str_struct.offset = READ32 (buf);
+		// this offset can be 64bit, based on dwarf format
+		if (hdr->is_64bit) {
+			value->encoding.str_struct.offset = READ64 (buf);
+		} else {
+			value->encoding.str_struct.offset = READ32 (buf);
+		}
 		if (debug_str && value->encoding.str_struct.offset < debug_str_len) {
 			value->encoding.str_struct.string = strdup (
 				(const char *)(debug_str +
 					value->encoding.str_struct.offset));
+			printf("%s\n", value->encoding.str_struct.string);
 		} else {
 			value->encoding.str_struct.string = NULL;
-		}
-		break;
-	case DW_FORM_udata:
-		{
-			ut64 ndata = 0;
-			const ut8 *data = (const ut8*)&ndata;
-			buf = r_uleb128 (buf, R_MIN (sizeof (data), (size_t)(buf_end - buf)), &ndata);
-			memcpy (&value->encoding.data, data, sizeof (value->encoding.data));
-			value->encoding.str_struct.string = NULL;
+			printf("NULL\n");
 		}
 		break;
 	case DW_FORM_ref_addr:
-		value->encoding.reference = READ64 (buf); // addr size of machine
+	// This is 4 or 8 bytes depending where it refers to
+	// http://www.dwarfstd.org/doc/Dwarf3.pdf page 128
+	/*
+	 For references from one shared object or
+	static executable file to another, the relocation and identification of the target object must be
+	performed by the consumer. In the 32-bit DWARF format, this offset is a 4-byte unsigned
+	value; in the 64-bit DWARF format, it is an 8-byte unsigned value
+	*/
+		// if (hdr->is_64bit) {
```

Comment:
Why this part is commented?

Reply:
Because I am unsure if the 4 or 8 bytes depends on the current compilation to be 32/64 bit or on the referenced compilation unit to be 32/64 bit, and the commented implementation is working for the first case.
Because this part
```
identification of the target object must be performed by the consumer.
```
Seems like I need to indentify the referenced CU and its bit format